pkg/cache/internal/transformers.go

```diff
@@ -0,0 +1,49 @@
+package internal
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type TransformFuncByObject struct {
+	structured   TransformFuncByGVK
+	unstructured TransformFuncByGVK
+	metadata     TransformFuncByGVK
+}
+type TransformFuncByGVK map[schema.GroupVersionKind]cache.TransformFunc
+
+func NewTransformFuncByObject() *TransformFuncByObject {
+	return &TransformFuncByObject{
+		structured:   TransformFuncByGVK{},
+		unstructured: TransformFuncByGVK{},
+		metadata:     TransformFuncByGVK{},
+	}
+}
+
+func (t TransformFuncByObject) Set(obj runtime.Object, scheme *runtime.Scheme, transformer cache.TransformFunc) error {
+	gvk, err := apiutil.GVKForObject(obj, scheme)
+	if err != nil {
+		return err
+	}
+
+	switch obj.(type) {
```

Comment:
So you are ram pressed, you watch the same things in two different representations and in order to make that more efficient, you strip the data you already have in the metadata representation from the full representation...? Can you explain why you don't use the full representation throughout instead?

fwiw, `TransformFunc` is defined as  `func(interface{}) (interface{}, error)` and doesn't even godoc what you should expect to get except for `cache.DeletedFinalStateUnknown` - I personally would need to read the surrounding code or trial and error to figure out how to use it correctly one way or another.

Reply:
Apologies, I misrepresented/remembered the use case. You are right that it would be so simple to use the structured object the way I explained it!

In the case of the metadata watches, only the partial metadata is actually watched so in fact your proposal to split by only GVK is not an issue for this case. I was mistaken.

To your other comment: I've taken a note to improve the documentation/type requirements of the TransformFunc upstream in client-go

Reply:
Given the above, will wait a tiny bit to see if there is any more feedback on this API design before changing it to use a single transform for all representations of a GVK.

Reply:
> If someone writes a transform that expects a specific representation and then creates a cache for the same type in a different representation, that is effectively a bug.

This may happen if someone creates duplicate caches for the same GVK, as https://github.com/kubernetes-sigs/controller-runtime/issues/1660 recorded. But turns out only a comment doc has been added.
Should we have some mechanisms to check and warn users? Otherwise the one `Transform` for one GVK may work unexpectedly and have risk of panic.

Reply:
I've changed the API to use one transform for each representation. Still need to update tests/polish. Marking this one as resolved.